Introduce callbacks for Task lifecycle

paarthmadan opened the pull request:
Closes #334 

#### Motivation

Having callbacks provides the client the opportunity to integrate with the task lifecycle. As cited in the original issue, a Slack notification (or any other message dispatch) seems to be a common use case when running tasks such as these.

#### What was introduced?

This PR adds an interface for supporting callbacks in `MaintenanceTasks::Task`. 

It provides the following public class methods: 

- `after_start`
- `after_pause`
- `after_interrupt`
- `after_error`
- `after_cancel`
- `after_complete`

One important behaviour is that errors that are raised during the custom error callback code are **ignored / swallowed**, to guarantee that our `error_handler` proceeds as expected.

Internally, the events added to the object lifecycle are: `start`, `complete`, `pause,` `interrupt`, `cancel`, `error`.

The `TaskJobConcern` is responsible for running the task's callbacks and they correspond with the lifecycle of the `TaskJob` itself.

The PR updates documentation for the new feature. I've made sure to add a note about how a client might share callback behaviour across Tasks inspired by [this comment](https://github.com/Shopify/maintenance_tasks/pull/484#issuecomment-918313190) and this [discussion](https://github.com/Shopify/maintenance_tasks/pull/77#discussion_r508941116).

#### Testing

I've introduced a new maintenance task in the dummy app called `CallbackTestTask`. This task uses the new callback hooks. 

In the test, I assert that these methods are called appropriate by the `TaskJobConcern` using Mocha expectations.

etiennebarrie reviewed (commented):
I'm not sure about the names `on_task_start`/`on_task_complete`. While it’s used by job-iteration, it doesn't feel idiomatic in Rails.
I’ve started investigating, but I’d love it if you want to take a stab at looking at existing callbacks and finding better names.

We will need a README section for this too.

I wasn’t sure about setting it on Task, since we don’t have an abstract class for all application tasks, and users don’t want to have set the same callbacks on all their task for Slack updates for example, but then I realized we can always set some callbacks on MaintenanceTasks::Task from the initializer. (Although, we have to think about autoloading in that case).

paarthmadan commented on app/tasks/maintenance_tasks/task.rb:
```diff
@@ -104,6 +106,22 @@ def throttle_on(backoff: 30.seconds, &condition)
+      # @param filter_list apply filters to the callback
+      #   (see https://api.rubyonrails.org/classes/ActiveSupport/Callbacks/ClassMethods.html#method-i-set_callback)
+      def on_task_start(*filter_list, &block)
+        set_callback(:start, :after, *filter_list, &block)
```
No specific reason, simply mimicked how callbacks were setup in `job-iteration`. Do you think it makes sense to leave this to the client?

etiennebarrie commented:
> I had a similar hesitance towards introducing the interface on the `Task`. Do you think it makes sense to introduce an abstract class?

We had one and we removed it: https://github.com/Shopify/maintenance_tasks/pull/124

Since most configuration goes on the MaintenanceTasks module or in the custom job, I'm not sure we need to introduce an abstract class just for this.

The autoloading problem could be solved by documentation, (mention `Rails.autoloaders.main.on_load("MaintenanceTasks::Task")`) or we could store the configuration on MaintenanceTask, and when Task is loaded, it could pull the callback configuration from there.

etiennebarrie reviewed (commented):
We need to handle exceptions in callbacks.
Maybe we can let them bubble and be handled by the error handler, but even in that case we have to handle an exception in the error callback, otherwise it prevents the error handler from running.

paarthmadan commented on README.md:
```diff
@@ -235,6 +235,41 @@ to run. Since arguments are specified in the user interface via text area
+
+`on_start`
+`on_pause`
+`on_interrupt`
```
Right, I too wasn't able to identify a particularly strong use case. My rationale is similar to your counter-argument; we want to enable clients to integrate with the lifecycle of the task and one of the states of a task is `interrupted`.

paarthmadan commented on README.md:
````diff
@@ -235,6 +235,41 @@ to run. Since arguments are specified in the user interface via text area
+end
+```
+
+To share callback behaviour across all Tasks, you can use a module. If you choose to include the module on `MaintenanceTasks::Task` in an initializer, you should use `on_load`
````
Great callout, I'll rework this section 👍

gmcgibbon commented on README.md:
```diff
@@ -235,6 +235,41 @@ to run. Since arguments are specified in the user interface via text area
+`on_interrupt`
+`on_error`
+`on_cancel`
+`on_complete`
```
The `on_` sounds too much like JavaScript to me. I would vote for `before_*`/`after_*`. See https://api.rubyonrails.org/classes/ActiveSupport/Callbacks.html, this way you can keep the callback names the same as well (eg. `before_error`, `after_error`).

paarthmadan replied:
Good callout. 

After doing some digging in Rails source, I've found a few cohorts for naming callback hooks:

1. `before_`, `after_` and occasionally `around_`. This seems to be the most frequently appearing. An example: [ActiveRecord callbacks.](https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-around_save)
2. `to_`. Examples: [Execution Wrapper](https://api.rubyonrails.org/classes/ActiveSupport/ExecutionWrapper.html#method-c-to_complete) (`to_run`, `to_complete`), [Reloader](https://api.rubyonrails.org/classes/ActiveSupport/Reloader.html#method-c-to_prepare) (`to_prepare`)
3. Past Participle. Note, for the example you cited `resets` is aliased as `after_reset`, too.

I think we should consider either (1) or (3). I'm leaning towards (1) and naming them `after_start`, `after_interrupt`, `after_error`, etc. We are in fact registering a callback _after_ in most cases, so I suppose this name is most true or correct.

I'm impartial, so I'm open to including a few others in the discussion and hearing their opinions. Once we get some alignment, I'll make the name change 👍

gmcgibbon commented on README.md:
````diff
@@ -235,6 +235,41 @@ to run. Since arguments are specified in the user interface via text area
+To share callback behaviour across all Tasks, you can use a module. If you choose to include the module on `MaintenanceTasks::Task` in an initializer, you should use `on_load`
+
+```ruby
+Rails.autoloaders.main.on_load("MaintenanceTasks::Task") do
````
This is lazier than to_prepare though. Any to_prepare hook will run on each boot which could slow down an app. Since the Zeitwerk autoloader is default in Rails now we can assume most people will have it.

paarthmadan replied:
> This is lazier than to_prepare though. 

@gmcgibbon I'm assuming you meant lazier than `on_load` (but please correct me if I've misinterpreted)

Do you recommend instead the initial suggestion then?

```ruby
Rails.autoloaders.main.on_load("MaintenanceTasks::Task") do
  MaintenanceTasks::Task.after_start :notify
end
```

etiennebarrie commented on app/jobs/concerns/maintenance_tasks/task_job_concern.rb:
```diff
@@ -162,7 +169,11 @@ def on_error(error)
+      @task.run_callbacks(:error) if defined?(@task)
       errored_element = @errored_element if defined?(@errored_element)
+    rescue StandardError => e
+      MaintenanceTasks.error_handler.call(e, task_context, errored_element)
```
I think we'll want to wrap the call to run_callbacks only, not the whole method

paarthmadan commented on README.md:
````diff
@@ -235,6 +235,67 @@ to run. Since arguments are specified in the user interface via text area
+```
+
+All other callbacks will bubble up their exceptions,
+and eventually handled by the [error handler](customizing-the-error-handler).
````
This was a discrepancy, thanks for the callout.

paarthmadan commented:
Thanks for the reviews. I've removed the extraneous call to the error handler. Instead, I isolate the rescuing of `StandardError` to just the `run_callbacks(:error)` method, and re-raise a new `ErrorCallbackError` on error. This new error is rescued and handled (by doing nothing) in the main procedure.

I've also addressed documentation / naming concerns.

I'll rewrite commit history so it makes more sense before merging 👍 

cc: @etiennebarrie @gmcgibbon

paarthmadan commented on README.md:
````diff
@@ -235,6 +235,69 @@ to run. Since arguments are specified in the user interface via text area
+```ruby
+# config/initializer/maintenance_tasks.rb
+Rails.autoloaders.main.on_load("MaintenanceTasks::Task") do
+  MaintenanceTasks::Task.class_eval do
````
In fact, this _was_ Gannon's suggestion. See: https://github.com/Shopify/maintenance_tasks/pull/484#discussion_r712385169. I'm happy to change this to a block unless there's a clear disadvantage to doing so.
